collector.go

```diff
@@ -112,6 +122,14 @@ func (c *collector) Collect(ch chan<- prometheus.Metric) {
 		return
 	}
 
+	consumers, err := client.Consumergroups()
+	if err != nil {
+		msg := fmt.Sprintf("Error collecting list of consumers: %s", err)
+		log.Error(msg)
+		ch <- prometheus.NewInvalidMetric(prometheus.NewDesc("zookeeper_consumer_list_error", msg, nil, nil), err)
```

Comment:
Sorry, but this is- and the other InvalidMetric usage- is an antipattern.  If this occurs you wind up with partial scrape data (or none) and have to infer the failure via absence checks- which isn't guaranteed.

The standard approach here is to export an atomically incrementing error count for things like this- that surfaces the failure in a way that a query can track it (via `delta()`) rather than just going partially radio silent.

Reply:
The NewInvalidMetric propagates the full text of the error back through the channel, so it ends up both on stderr and in the body of the document returned to the client. It's quite useful for debugging, just not monitoring. I think it would be more precise to say that it's incomplete, rather than wrong.

I'd prefer to only review one change at a time, so I'm going to implement a zookeeper error metric in a subsequent PR, unless there's something about this implementation that adds a new bug that we don't already have.

Reply:
Respectfully, you're adding a new metric here- which means you have a responsibility to implement it properly and to prometheus standards.  I pointed this flaw out in your last PR with the tacit understanding that you'd do a followup and fix the antipattern- you're instead trying to expand it further, and there is zero reason to do so- you're actively removing information from scraping and that's not how exporters are implemented.

It'll cost more time debating "can I do it in the next PR" then just implementing an error gauge- it's a 2-3 minute thing.  Please do so.

Reply:
This was the second PR that we didn't get to at that review. The only thing I'm resisting is reordering the patches. Unless I'm missing something, this doesn't make anything worse, because any 3rd-party scraper that chokes on this would already be choking.

Reply:
Here's the follow-on that fixes incomplete scrape issues. I'll submit it as soon as this is merged, but reordering these patches would take substantially longer than 2-3 minutes.

https://github.com/csnook/kafka_zookeeper_exporter/commit/c64b29e6246f428f77be5e6677a3c6ee8f439282

Reply:
After a fix for proper use of atomics, that commit is here: https://github.com/csnook/kafka_zookeeper_exporter/commit/2e724cdeca59946a92e345736b9b9c819cb211c9

Reply:
After fixing the use of atomics, that commit is here: https://github.com/csnook/kafka_zookeeper_exporter/commit/eeff3f82fb54980bc4ec0ddf49a16183c142647a

Reply:
Please include the fixup commits into this PR- it'll centralize the review, and approval of this PR is dependent on those fixups going in at the same time.

Additionally, I've commented w/in those commits that you're not implementing a per scrape metric- you hardwired it so one bad scrape persists the failure value through all.

Reply:
Done